src/misc/cloudAuth.ts

```diff
@@ -0,0 +1,194 @@
+import { Request, Server } from '@hapi/hapi';
+import axios from 'axios';
+import Conf = require('conf');
+import { createHash } from 'crypto';
+import open = require('open');
+import { stringify } from 'querystring';
+import { cpu, mem, osInfo, system } from 'systeminformation';
+import { v4 as uuidv4 } from 'uuid';
+
+const cloudUrl = 'https://cloud-beta.rocket.chat';
+const clientId = '5d8e59c5d48080ef5497e522';
+const scope = 'offline_access marketplace';
+
+export interface ICloudToken {
+    access_token: string;
+    expires_in: number;
+    scope: string;
+    refresh_token: string;
+    token_type: string;
+}
+
+export interface ICloudAuthStorage {
+    token: ICloudToken;
+    expiresAt: Date;
+}
+
+export class CloudAuth {
+    private config: Conf;
+    private codeVerifier: string;
+    private port = 3005;
+    private server: Server;
+    private redirectUri: string;
+
+    constructor() {
+        this.redirectUri = `http://localhost:${ this.port }/callback`;
+        this.codeVerifier = uuidv4() + uuidv4();
+    }
+
+    public async executeAuthFlow(): Promise<string> {
+        await this.initialize();
+
+        return new Promise((resolve, reject) => {
+            try {
+                this.server = new Server({ host: 'localhost', port: this.port });
+                this.server.route({
+                    method: 'GET',
+                    path: '/callback',
+                    handler: async (request: Request) => {
+                        try {
+                            const code = request.query.code;
+                            const token = await this.fetchToken(code);
+
+                            resolve(token.access_token);
+                            return 'Thank you. You can close this tab.';
+                        } catch (err) {
+                            reject(err);
+                        } finally {
+                            this.server.stop();
+                        }
+                    },
+                });
+
+                const codeChallenge = createHash('sha256').update(this.codeVerifier).digest('base64');
+                const authorizeUrl = this.buildAuthorizeUrl(codeChallenge);
+                open(authorizeUrl);
```

Comment:
Maybe we should print it out to the command line to?  Something like:

If your browser didn't open please follow the link below:
<link>